### ISSUE TITLE
Cancel in progress deal that is PRE-publication

### ISSUE THREAD

**Issue by jsign** [filecoin-project/lotus#4093]
It would be useful to signal a cancelation of an ongoing deal since from the user's perspective some deals might be stuck for a long time and the same content can't be sent again to the same miner with a new deal proposal.

Even if retrying isn't the original intention, might be useful to cancel a deal for other reasons.

I'd expect that it might have some consequences depending on the stage in which a user wants to cancel the deal since if the miner is already sealing there might be some tension between _accepting_ that cancellation or not.

*edit post-publication cancel requires on chain changes and will be rejected out of hand for now*

Planting the seed for discussion.

**Comment by jsign** [comment]
So the above was more about a "clean" cancellation of a fired deal.

There's another case which we're facing now, and is that the miner stopped dealing with a particular deal but from the client Lotus state it seems it still sealing. So it might be also useful to completely cancel or abandon "locally" those deals to get unstuck and retry again.

To be clear, the above sounds like a workaround for some bug, and in some way implies some risk that the to-be-abandoned deal is still being executed by the miner and would eventually land on-chain (thus, incurring in payment for the client). Saying it differently, this kind of "forced cancelation" doesn't guarantee that deal won't land on chain.

**Comment by jennijuju** [comment]
@jsign are you suggesting this from miner's perspective? And do you mean you want to be able to cancel a deal from lotus CLI?

**Comment by jsign** [comment]
I'm talking from the users perspective.

The main thing now is: "If I made a deal with a miner, and the miner has abandoned/errored/etc that deal. I want to have a way to get unstuck from my side and just consider that deal "failed/canceled""

**Comment by jennijuju** [comment]
@jsign yep, its the case for both sides. Will have the team to take a look.

**Comment by pooja** [comment]
If a miner wants to cancel an in-progress deal, can that happen?

**Comment by hannahhoward** [comment]
follow-up discussion on this:
- data transfers automatically self cancel after an hour long TCP timeout
- we have manual commands to cancel data transfers on client and miner
- cancellation POST publish is extremely difficult -- it requires a consensus breaking change to allow clients to cancel their deal without penalty to the miner -- I do not see us implementing this any time soon
- I do think it would be useful to have a prepublication cancel for deals
- I am editing this ticket to limit scope to prepublication

Here is what is needed to implement it:

- The StorageClient in go-fil-markets needs a Cancel command
- The command should error in the case of the current client state being anything post publish
- We should modify the deal proposal protocol to accept a cancel attribute (need to do the work to insure successful protocol upgrade and support of legacy protocols)
- The storage client cancel command should send a libp2p request to cancel the deal to the provider
   - if the provider does not support the latest protocol version, the cancel should fail
   - in the provider, when processing a cancel, if the storage deal is in StorageDealPublishing or after, the cancel should fail and a rejection should be sent
   - if the provider is in an acceptable state, dispatch a cancel event to update provider state to cancelled & respond with an acceptance of cancellation
   - if the provider rejects, the Cancel command should fail
- If the provider accepts cancel, the client should dispatch an event to update its state to cancelled and return success
- Probably need to also account for cancelling data transfer as well

- need corresponding PR in Lotus for api updates and cli

- This is a lot of work still and I'm leaving it P2

**Comment by jsign** [comment]
@hannahhoward , thanks for de detailed answer.
I agree that post-cancelation is a very hard problem and, actually, wasn't part of my request since the miner has little incentives to pay the cost of whatever it planned when accepting the deal. Canceling pre-publish deals sounds good, but I wouldn't consider that urgent.

My main point for this PR was about canceling a deal only on the client-side, even if that means that deal would get published and continue by its own means. This request was mostly a workaround of having tons of deals stuck in _Sealing_ or other statuses. In short, cleaning tracked deals only client-side.

Most of these "getting stuck" situations are more formal issues now, e.g:https://github.com/filecoin-project/lotus/issues/4990

Also, as you confirmed yesterday, once you get a very slow miner on your sending queue, the client gets in a hairy situation. Those deals are affecting total throughput. As we know, canceling the transfers doesn't [fully solve](https://github.com/filecoin-project/lotus/issues/4997) the problem.

Even if you don't have a clean pre-publishing cancelation, just a hard cancelation only client-side would be enough to forget about all those deals and move on as a generic solution of any "getting stuck situation", or just saying "I want to start fresh again without starting a new Lotus node".

**Comment by dineshshenoy** [comment]
Current behavior is that the deal will expire after epoch start date.  So background noise in the interim.  @jsign - is this still a priority?

**Comment by jsign** [comment]
The motivating cause of this was canceling deals that weren't in a final status in the client-side (accepting the risk that the deal will be on-chain).

If a deal is still on data transfer, but for whatever reason is stuck there, I think it would be useful to hard-cancel it and delete it.  Same with maybe other statuses.

I think this is a useful feature to avoid having long dangling deals.

**Comment by jennijuju** [comment]
> follow-up discussion on this:
> 
> * data transfers automatically self cancel after an hour long TCP timeout
> * we have manual commands to cancel data transfers on client and miner
> * cancellation POST publish is extremely difficult -- it requires a consensus breaking change to allow clients to cancel their deal without penalty to the miner -- I do not see us implementing this any time soon
> * I do think it would be useful to have a prepublication cancel for deals
> * I am editing this ticket to limit scope to prepublication
> 
> Here is what is needed to implement it:
> 
> * The StorageClient in go-fil-markets needs a Cancel command
> * The command should error in the case of the current client state being anything post publish
> * We should modify the deal proposal protocol to accept a cancel attribute (need to do the work to insure successful protocol upgrade and support of legacy protocols)
> * The storage client cancel command should send a libp2p request to cancel the deal to the provider
>   
>   * if the provider does not support the latest protocol version, the cancel should fail
>   * in the provider, when processing a cancel, if the storage deal is in StorageDealPublishing or after, the cancel should fail and a rejection should be sent
>   * if the provider is in an acceptable state, dispatch a cancel event to update provider state to cancelled & respond with an acceptance of cancellation
>   * if the provider rejects, the Cancel command should fail
> * If the provider accepts cancel, the client should dispatch an event to update its state to cancelled and return success
> * Probably need to also account for cancelling data transfer as well
> * need corresponding PR in Lotus for api updates and cli
> * This is a lot of work still and I'm leaving it P2

> The StorageClient in go-fil-markets needs a Cancel command
Seems like lotus side of things are blocked by impl in go-fil-markets, re-add this ticket back to the lotus board when there is action items for lotus

**Comment by TippyFlitsUK** [comment]
Hi 👋 

The Legacy Lotus Markets sub-system [reached EOL](https://github.com/filecoin-project/lotus/releases/tag/v1.18.0) at the end of the 31st January 2023.

This ticket is being marked as `won't fix` and closed as the Lotus Team will no longer be making any further fixes or enhancements to the legacy markets subsystem.

Please feel free to re-open this ticket in the new Boost markets sub-system repository at https://github.com/filecoin-project/boost if you feel that it is still relevant.

Many thanks :pray:
